File: Largest_subarray_with_0_sum/largest_subarray_0_sum.cpp
#include<iostream>
#include<unordered_map>
#include<vector>
using namespace std;
// Function to find the largest subarray with sum 0
int largestSubarrayWithSumZero(vector<int>& arr) {
    unordered_map<int, int> sumIndexMap; // Map to store sum and its index
    int maxLength = 0; // Variable to store the maximum length of subarray
    int sum = 0; // Variable to store the cumulative sum

    for (int i = 0; i < arr.size(); i++) {
        sum += arr[i]; // Update the cumulative sum

        // If the cumulative sum is 0, update maxLength
        if (sum == 0) {
            maxLength = i + 1;
        }

        // If the cumulative sum has been seen before, update maxLength
        if (sumIndexMap.find(sum) != sumIndexMap.end()) {
            maxLength = max(maxLength, i - sumIndexMap[sum]);
        } else {
            // Store the index of the first occurrence of this sum
            sumIndexMap[sum] = i;
        }
    }
    return maxLength; // Return the maximum length of subarray with sum 0
}
// Main function to test the largestSubarrayWithSumZero function
int main() {
    vector<int> arr = {1, -1, 2, -2, 3, -3, 4}; // Example array
    int result = largestSubarrayWithSumZero(arr); // Call the function
    cout << "The length of the largest subarray with sum 0 is: " << result << endl; // Output the result
    return 0; // Return success
}
// Output: The length of the largest subarray with sum 0 is: 6

File: Largest_subarray_with_0_sum/largest_subarray_0_sum.rs
use::std::collections::HashMap;
use::std::cmp::max;
use::std::vec::Vec;
use::std::i32;
fn largest_subarray_with_0_sum(arr: Vec<i32>) -> i32 {
    let mut sum_map = HashMap::new();
    let mut max_length = 0;
    let mut sum = 0;

    for (i, &num) in arr.iter().enumerate() {
        sum += num;

        if sum == 0 {
            max_length = max(max_length, (i + 1) as i32);
        }

        if let Some(&prev_index) = sum_map.get(&sum) {
            max_length = max(max_length, (i - prev_index) as i32);
        } else {
            sum_map.insert(sum, i);
        }
    }

    max_length
}
fn main() {
    let arr = vec![1, -1, 2, -2, 3, -3, 4, -4];
    let result = largest_subarray_with_0_sum(arr);
    println!("Length of the largest subarray with sum 0: {}", result);
}